enhancements/arbiter-clusters.md

```diff
@@ -0,0 +1,628 @@
+specified.
+
+`installConfig.yaml`
+
+```yaml
+apiVersion: v1
+baseDomain: devcluster.openshift.com
+compute:
+  - architecture: amd64
+    hyperthreading: Enabled
+    name: worker
+    platform: {}
+    replicas: 0
+arbiterNode:
+  architecture: amd64
+  hyperthreading: Enabled
+  replicas: 1
+  name: arbiter
+  platform:
+    aws:
+      type: m6a.large
+controlPlane:
+  architecture: amd64
+  hyperthreading: Enabled
+  name: master
+  platform:
+    aws:
+      type: m6a.4xlarge
+  replicas: 2
+```
+
+Ex. `/machines/arbiter.go`
+
+```go
+type Arbiter struct {
+	UserDataFile           *asset.File
+	MachineConfigFiles     []*asset.File
+	MachineFiles           []*asset.File
+	ControlPlaneMachineSet *asset.File
+	IPClaimFiles           []*asset.File
+	IPAddrFiles            []*asset.File
+
+	// SecretFiles is used by the baremetal platform to register the
+	// credential information for communicating with management
+	// controllers on hosts.
+	SecretFiles []*asset.File
+
+	// NetworkConfigSecretFiles is used by the baremetal platform to
+	// store the networking configuration per host
+	NetworkConfigSecretFiles []*asset.File
+
+	// HostFiles is the list of baremetal hosts provided in the
+	// installer configuration.
+	HostFiles []*asset.File
+}
+```
+
+### MCO Changes
+
+Currently MCO blocks machine config pools that target non master/worker roles in
+a [validation
+webhook](https://github.com/openshift/machine-config-operator/blob/7c5ae75515fe373e3eecee711c18b07d01b5e759/manifests/machineconfigcontroller/custom-machine-config-pool-selector-validatingadmissionpolicy.yaml#L20-L25).
+We will need to update this to allow the new arbiter role.
+
+`custom-machine-config-pool-selector-validatingadmissionpolicy.yaml`
+
+```
+          has(object.spec.machineConfigSelector.matchLabels) &&
+          (
+            (object.spec.machineConfigSelector.matchLabels["machineconfiguration.openshift.io/role"] == "master")
+            ||
+            (object.spec.machineConfigSelector.matchLabels["machineconfiguration.openshift.io/role"] == "worker")
+            ||
+            (object.spec.machineConfigSelector.matchLabels["machineconfiguration.openshift.io/role"] == "arbiter")
+          )
+```
+
+We will also need to update the go logic in the renderer to allow the `arbiter`
+role.
+
+`render.go`
+
+```go
+// GenerateMachineConfigsForRole creates MachineConfigs for the role provided
+func GenerateMachineConfigsForRole(config *RenderConfig, role, templateDir string) ([]*mcfgv1.MachineConfig, error) {
+	rolePath := role
+	//nolint:goconst
+	if role != "worker" && role != "master" && role != "arbiter" {
+		// custom pools are only allowed to be worker's children
+		// and can reuse the worker templates
+		rolePath = "worker"
+	}
+...
+```
+
+We then need to add new rendered initial files for the arbiter role so
+bootstrapping occurs correctly. When MCO stands up the Machine Config Server
```

Comment:
As a note this also creates a set of arbiter-user-data and arbiter-user-data-managed secrets in the mapi namespace. I don't suppose we'd want to scale additional arbiter nodes post-install though, so those would just be unused (which is probably fine?) but the user would theoretically be able to leverage them to scale additional arbiter nodes.

Also to check, do we foresee anyone adding arbiter nodes post-install? Maybe we can just have the bootstrap MCS server arbiter and block the in-cluster MCS from doing so.

Reply:
At the time being there is no need to scale the arbiter node, but there is no reason why we can't do that in the future for small HA deployments. In the installer changes I left the replica field to allow us to revisit this in the future, we can put in a validation check for now to enforce 1 arbiter. However, I can imagine a future scenario where we configure a 2 Arbiter 1 Master.